log/src/io/pedestal/log.clj

```diff
@@ -122,24 +131,94 @@
+                             resolve))
+
+(def log-level-dispatch
+  {:trace -trace
+   :debug -debug
+   :info  -info
+   :warn  -warn
+   :error -error})
+
+(defn log
+  "This function provides basic/core logging functionality as a function.
+  You may prefer to use this if you need custom logging functionality beyond
+  what is offered by the standard Pedestal logging marcos (which in turn just call the protocols).
+
+  Given a map of logging information,
+    and optionally a default log-level keyword (if not found in the map) -- default is :info,
+  Determine if the appropriate logger to use,
+   determine if logging-level is enabled,
+   format the logging message,
+  And return the result of calling the appropriate logging function, dispatched to the logging protocols.
+
+  Special keys within the log message:
+   :level -- A keyword, the log level to use for this message, defaults to `default-level`
+   :exception -- A Throwable/Exception to log
+   :io.pedestal.log/logger -- The logger to use for this message,
+                              defaults to the `override-logger` or the SLF4J logger
+   :io.pedestal.log/logger-name -- A String, the loggerName to use if SLF4J logger is used,
+                                   defaults to `*ns*` which may be 'clojure.core' depending on execution,
+   :io.pedestal.log/formatter -- A single-arg function that when given a map, returns a String for logging,
+                                 defaults to `pr-str`
+
+  If using this function within a macro, you're encouraged to merge all 'meta' information
+  (like line info) into the log message map.
+  For example:
+
+  (defmacro log-macro [log-map]
+  (let [named-log-map (if (::logger-name log-map)
+                        log-map
+                        (assoc log-map ::logger-name (name (ns-name *ns*))))
+        final-log-map (assoc named-log-map :line (:line (meta &form)))]
+    `(log ~final-log-map :info)))
+  "
+  ([keyvals]
+   (log keyvals :info))
+  ([keyvals default-level]
+   (let [keyvals-map (if (map? keyvals) keyvals (apply array-map keyvals))
+         level (:level keyvals-map default-level)
+         logger (or (::logger keyvals-map)
+                    override-logger
+                    (LoggerFactory/getLogger ^String (or (::logger-name keyvals-map)
+                                                         (name (ns-name *ns*)))))]
+     (when (io.pedestal.log/-level-enabled? logger level)
+       (let [exception (:exception keyvals-map)
+             formatter (::formatter keyvals-map pr-str)
+             ;; You/Users have full control over binding *print-length*, use it wisely please
+             msg (formatter (dissoc keyvals-map
+                                    :exception ::logger ::logger-name ::formatter :level))
+             ;; In order to get to here, `level` has to be enabled,
+             ;;  so it should be safe to look-up in the dispatch
+             log-fn (get log-level-dispatch level)]
```

Comment:
This is the exact comment you just argued against :)  There's a check at the top (if log the level is enabled), so we know that the log level is valid and it's enabled when this line runs.  I originally had `identity` as the not-found value, but I'd prefer to see NPE happen, since it means the whole world has fallen apart.

Reply:
Optionally, we could accept a custom `log-fn` as an option in the map.

Reply:
Sorry, I missed the call to `-level-enabled?` :). We already accept an override logger so does accepting a custom log fn make sense?

Reply:
I personally don't think it makes sense.  I'd prefer to keep people focused on the protocols instead.